### PR TITLE
#1333 kraken to helix

### PR DESCRIPTION
**Pull request by CaveMobster**
### Description of the Change
This PR does the following:
- Updates the twitch API package
- Changes all Kraken API calls to Helix calls 
- Cleans up Helix calls like paginated block lists and stream tags
- Updates several Helix calls that were not able to use the package methods yet
- Moves and reworks the chat role functionality to a separate chat roles manager, since Helix doesn't offer a way to get chat roles
- Moves all team roles logic to the team roles manager, so that all team role information will be accessed through there
- Fixes a bug in the active viewer list effect
- Removes the teams api, the channel progression restriction and the channel access

### Applicable Issues
https://github.com/crowbartools/Firebot/issues/1333

### Testing
Tested all functionality that was touched and updated, and checked for backwards compatibility.

### CODE REVIEW

File: backend/roles/chat-roles-manager.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+        if (moderators) {
+            for (const moderator of moderators) {
+                if (moderator.userId === userId) {
+                    return true;
+                }
+            }
+        }
+
+        return false;
```
Comment (ebiggz) [inline backend/roles/chat-roles-manager.js:27]: we can shorten this to:
```js
return moderators ? moderators.some(m => m.userId === userId) : false;
```

File: backend/roles/chat-roles-manager.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+        if (vips) {
+            for (const vip of vips) {
+                if (vip === username) {
+                    return true;
+                }
+            }
+        }
+
+        return false;
```
Comment (ebiggz) [inline backend/roles/chat-roles-manager.js:47]: We can shorten this similarly to the mods suggestion:
```js
return vips ? vips.includes(username) : false;
```

File: backend/roles/chat-roles-manager.js [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+            roles.push("mod");
+        }
+
+        if (userIsVip) {
+            roles.push("vip");
+        }
+
+        if (subRoles) {
+            roles = roles.concat(subRoles);
+        }
```
Comment (ebiggz) [inline backend/roles/chat-roles-manager.js:123]: subRoles will never be null (it returns array or empty array) so this check isn't needed.

File: backend/roles/team-roles-manager.js [hunk 4]
```diff
@@ -2,45 +2,62 @@
...
-                id: team.id,
+                id: parseInt(team.id), // For compatibility (Kraken API returned an int)
                 name: team.displayName
             };
         });
-}
-
-async function getAllTeamRolesForViewer(username) {
-    const roles = await twitchApi.teams.getMatchingTeamsByName(username);
+};
```
Comment (ebiggz) [inline backend/roles/team-roles-manager.js:22]: Great thinking on making sure this remains compatible. 

Super minor, but we can strip away some unneeded syntax like so
```js
const mapRoles = (teams) => teams.map(t => ({
  id: parseInt(t.id),
  name: t.displayName
})
```

File: backend/roles/team-roles-manager.js [hunk 5]
```diff
@@ -2,45 +2,62 @@
...
-}
+/** @returns {Promise.<MappedTeamRole[]>} */
+const getAllTeamRolesForViewer = async (username) => {
+    const client = twitchApi.getClient();
+    const user = await client.helix.users.getUserByName(username);
+    const streamerTeams = await client.helix.teams.getTeamsForBroadcaster(accountAccess.getAccounts().streamer.userId);
 
-async function getTeamRoles() {
-    const teams = await twitchApi.teams.getStreamerTeams();
+    if (streamerTeams == null) return null;
```
Comment (ebiggz) [inline backend/roles/team-roles-manager.js:30]: Might make sense to just return an empty array here instead of null so any code that calls this function doesn't have to worry about null checks. That's typically my rule of thumb with functions that returns arrays.

File: backend/roles/team-roles-manager.js [hunk 6]
```diff
@@ -2,45 +2,62 @@
...
-    if (teams == null) {
-        return null;
+    const teams = [];
+    for (const team of streamerTeams) {
+        const relations = await team.getUserRelations();
+        for (const relation of relations) {
+            if (relation.id === user.id) {
+                teams.push(team);
+            }
+        }
```
Comment (ebiggz) [inline backend/roles/team-roles-manager.js:39]: The outer `for` loop is unfortunately necessary because of the `await` call, but we could remove this nested one by instead doing:
```js
if(relations.some(r => r.id === user.id)) {
   teams.push(team);
}
```

File: backend/twitch-api/resource/stream-tags.js [hunk 7]
```diff
@@ -14,108 +12,55 @@ const { snakeKeys } = require('js-convert-case');
...
-            return null;
+        if (response) {
+            /**@type {TwitchStreamTag[]} */
+            return response.filter(tag => !tag.isAuto).map(tag => mapTwitchTag(tag));
         }
-
-        return response;
     } catch (error) {
         logger.error("Failed to get all stream tags", error);
         return null;
```
Comment (ebiggz) [inline backend/twitch-api/resource/stream-tags.js:35]: I know this is a previously exiting function. Just wanted to mention the returning an empty array instead of null thing again. Not sure how it would affect what is calling this though.

File: backend/twitch-api/resource/stream-tags.js [hunk 8]
```diff
@@ -14,108 +12,55 @@ const { snakeKeys } = require('js-convert-case');
...
-                "broadcaster_id": accountAccess.getAccounts().streamer.userId
-            },
-            body: snakeKeys(tagIds)
-        });
-        return true;
+        const client = twitchApi.getClient();
+        await client.helix.streams.replaceStreamTags(accountAccess.getAccounts().streamer.userId, tagIds);
     } catch (error) {
         logger.error("Failed to update channel stream tags", error);
         return false;
```
Comment (ebiggz) [inline backend/twitch-api/resource/stream-tags.js:60]: Returns false in the catch but the `return true` was removed from the try. Is the return boolean even used? If not, we can just remove the `return false` from the catch. If yes, we'll want to add the `return true` back to try.

File: backend/common/user-access.js [hunk 9]
```diff
@@ -84,17 +82,17 @@ async function getUserDetails(userId) {
 
     let isBlocked = false;
     try {
-        const blocklist = await twitchApi.users.getAllBlockedUsersPaginated(streamerData.userId);
+        const blocklist = await twitchApi.getClient().helix.users.getBlocksPaginated(streamerData.userId).getAll();
         if (blocklist) {
-            isBlocked = blocklist.find(id => id === twitchUser.id) != null;
+            isBlocked = blocklist.find(user => user.userId === twitchUser.id) != null;
```
Comment (ebiggz) [inline backend/common/user-access.js:87]: since we just want a boolean, instead of `.find` we can use `.some`:
```js
isBlocked = blocklist.some(u => u.userId === twitchUser.id);
```

**Review by ebiggz — CHANGES_REQUESTED** [review]

File: backend/roles/chat-roles-manager.js (hunk 1 above)
Comment (CaveMobster) [reply backend/roles/chat-roles-manager.js]: Fixed.

File: backend/roles/chat-roles-manager.js (hunk 2 above)
Comment (CaveMobster) [reply backend/roles/chat-roles-manager.js]: Fixed.

File: backend/roles/chat-roles-manager.js (hunk 3 above)
Comment (CaveMobster) [reply backend/roles/chat-roles-manager.js]: Fixed.

File: backend/roles/team-roles-manager.js (hunk 4 above)
Comment (CaveMobster) [reply backend/roles/team-roles-manager.js]: Fixed.

File: backend/roles/team-roles-manager.js (hunk 5 above)
Comment (CaveMobster) [reply backend/roles/team-roles-manager.js]: Fixed.

File: backend/roles/team-roles-manager.js (hunk 6 above)
Comment (CaveMobster) [reply backend/roles/team-roles-manager.js]: Fixed.

File: backend/twitch-api/resource/stream-tags.js (hunk 7 above)
Comment (CaveMobster) [reply backend/twitch-api/resource/stream-tags.js]: Fixed.

File: backend/twitch-api/resource/stream-tags.js (hunk 8 above)
Comment (CaveMobster) [reply backend/twitch-api/resource/stream-tags.js]: Fixed.

File: backend/common/user-access.js (hunk 9 above)
Comment (CaveMobster) [reply backend/common/user-access.js]: Fixed.
